src/graph/executor/algo/BFSShortestPathExecutor.cpp

```diff
@@ -1,52 +1,214 @@
+  auto iter = reverse ? ectx_->getResult(pathNode_->rightInputVar()).iter()
+                      : ectx_->getResult(pathNode_->leftInputVar()).iter();
   DCHECK(!!iter);
+  auto& visitedVids = reverse ? rightVisitedVids_ : leftVisitedVids_;
+  auto& allEdges = reverse ? allRightEdges_ : allLeftEdges_;
+  allEdges.emplace_back();
+  auto& currentEdges = allEdges.back();
+
+  auto iterSize = iter->size();
+  visitedVids.reserve(visitedVids.size() + iterSize);
+
+  std::unordered_set<Value> uniqueDst;
+  uniqueDst.reserve(iterSize);
+  DataSet nextStepVids;
+  nextStepVids.colNames = {nebula::kVid};
+  if (step_ == 1) {
+    for (; iter->valid(); iter->next()) {
+      auto edgeVal = iter->getEdge();
+      if (UNLIKELY(!edgeVal.isEdge())) {
+        continue;
+      }
+      auto& edge = edgeVal.getEdge();
+      auto dst = edge.dst;
+      visitedVids.emplace(edge.src);
+      if (uniqueDst.emplace(dst).second) {
+        nextStepVids.rows.emplace_back(Row({dst}));
+      }
+      currentEdges.emplace(std::move(dst), std::move(edge));
+    }
+  } else {
+    for (; iter->valid(); iter->next()) {
+      auto edgeVal = iter->getEdge();
+      if (UNLIKELY(!edgeVal.isEdge())) {
+        continue;
+      }
+      auto& edge = edgeVal.getEdge();
+      auto dst = edge.dst;
+      if (visitedVids.find(dst) != visitedVids.end()) {
+        continue;
+      }
+      if (uniqueDst.emplace(dst).second) {
+        nextStepVids.rows.emplace_back(Row({dst}));
+      }
+      currentEdges.emplace(std::move(dst), std::move(edge));
+    }
+  }
+  // set nextVid
+  const auto& nextVidVar = reverse ? pathNode_->rightVidVar() : pathNode_->leftVidVar();
+  ectx_->setResult(nextVidVar, ResultBuilder().value(std::move(nextStepVids)).build());
+
+  visitedVids.insert(std::make_move_iterator(uniqueDst.begin()),
+                     std::make_move_iterator(uniqueDst.end()));
+  return Status::OK();
+}
+
+folly::Future<Status> BFSShortestPathExecutor::conjunctPath() {
+  const auto& leftEdges = allLeftEdges_.back();
+  const auto& preRightEdges = allRightEdges_[step_ - 1];
+  std::vector<Value> meetVids;
+  bool oddStep = true;
+  for (const auto& edge : leftEdges) {
+    if (preRightEdges.find(edge.first) != preRightEdges.end()) {
+      meetVids.push_back(edge.first);
+    }
+  }
+  if (meetVids.empty() && step_ * 2 <= pathNode_->steps()) {
+    const auto& rightEdges = allRightEdges_.back();
+    for (const auto& edge : leftEdges) {
+      if (rightEdges.find(edge.first) != rightEdges.end()) {
+        meetVids.push_back(edge.first);
+        oddStep = false;
+      }
+    }
+  }
+  if (meetVids.empty()) {
+    return Status::OK();
+  }
+  size_t totalSize = meetVids.size();
+  size_t batchSize = totalSize / static_cast<size_t>(FLAGS_num_operator_threads);
+  std::vector<Value> batchVids;
+  batchVids.reserve(batchSize);
+  std::vector<folly::Future<DataSet>> futures;
+  for (size_t i = 0; i < totalSize; ++i) {
+    batchVids.push_back(meetVids[i]);
```

Comment:
If you iterate the meetVids vector from end to start, you will need not to copy the contents to a new vector.